xmppserver/src/main/java/org/jivesoftware/openfire/pubsub/PubsubNodeResultFilter.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (C) 2022 Ignite Realtime Foundation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jivesoftware.openfire.pubsub;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Comparator;
+import java.util.function.Predicate;
+
+/**
+ * Filters and sorts lists of pubsub nodes.
+ * <p>
+ * The class also supports pagination of results with the setStartIndex(int)
+ * and setNumResults(int) methods. If the start index is not set, it will
+ * begin at index 0 (the start of results). If the number of results is not set,
+ * it will be unbounded and return as many results as available.</p>
+ * <p>
+ * Factory methods to create common queries are provided for convenience.</p>
+ *
+ * @author Guus der Kinderen
+ * @see org.jivesoftware.openfire.SessionResultFilter
+ */
+public class PubsubNodeResultFilter
+{
+    private static final Logger Log = LoggerFactory.getLogger(PubsubNodeResultFilter.class);
+
+    // ############################################################
+    // Search order criteria
+    // ############################################################
+    /**
+     * Descending sort (ie 3, 2, 1...).
+     */
+    public static final int DESCENDING = 0;
```

Comment:
Wonder if this should be set to `ListPager.DESCENDING` ?

Reply:
(or even a `Boolean`, in line with my comments on `ListPager`

Reply:
I have been struggling with this a bit. On the other hand, it would be great to use the same field for all usages, but on the other hand, the two are also distinct, in the sense that the implementation of sorting cannot easily be generic: it is specific to the types of the particular element being processed.

Reply:
Perhaps pass in a `Comparable<T>` to the `ListPager` to provide the sorting?

Reply:
I liked this idea, but then ran into issues trying to implement it.

The `ListPager` doesn't really do any sorting. We could make it do that generically by providing a Comparator to replace the ugly integer-based arguments that we now have, but that introduces another issue. Those integer-based arguments are used to populate a hidden form, which is used to transfer the entire 'state' of sorting/filtering/pagination to the next request.

I've been looking at how to make this all a bit prettier, but after every change, something else keeps failing. I might be missing something obvious, but I'm getting a bit frustrated with fixing one thing, to find two new issues pop up. I'm thinking of merging this as-is, leaving improvements to future PRs.